Index: app/src/main/java/com/example/biomapper/RoiManager.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/app/src/main/java/com/example/biomapper/RoiManager.java b/app/src/main/java/com/example/biomapper/RoiManager.java
--- a/app/src/main/java/com/example/biomapper/RoiManager.java	(date 1646424391699)
+++ b/app/src/main/java/com/example/biomapper/RoiManager.java	(date 1646424391699)
@@ -124,9 +124,11 @@
         // Add a back button to the toolbar and make it functional.
         toolbar.setNavigationIcon(getContext().getDrawable(R.drawable.toolbar_back_icon));
         toolbar.setNavigationOnClickListener(
-                new View.OnClickListener() {
+                new View.OnClickListener()
+                {
                     @Override
-                    public void onClick(View view) {
+                    public void onClick(View view)
+                    {
                         openActionMenu();
                     }
                 }
@@ -151,7 +153,7 @@
             // Get indexes for removing/adding the map view.
             int oldMapIndex = roiParent.indexOfChild( mapView );
 
-            // Remove the view from the Main Map.
+            // Remove the view from the ROI Manger.
             roiParent.removeViewAt( oldMapIndex );
 
             // Create parameters for adding the view.
@@ -162,7 +164,7 @@
             viewParams.topMargin = 0;
             viewParams.bottomMargin = 0;
 
-            // Add the Base Map view to the Roi Manager
+            // Add the Base Map view to the Main Map.
             mainMapParent.addView( mapView, 0, viewParams );
         }
 
